test/gateway/eip20_gateway/progress_revert_stake.js

```diff
@@ -0,0 +1,456 @@
+    let burner = NullAddress;
+
+    bountyAmount = new BN(revokedProofData.gateway.constructor.bounty);
+
+    gateway = await Gateway.new(
+      mockToken.address,
+      baseToken.address,
+      coreAddress,
+      bountyAmount,
+      organization.address,
+      burner,
+    );
+
+    gatewayLib = await GatewayLib.deployed();
+
+    let paramsData = revokedProofData.gateway.stake.params;
+
+    revertStakeParams = {
+      amount: new BN(paramsData.amount, 16),
+      beneficiary: paramsData.beneficiary,
+      gasPrice: new BN(paramsData.gasPrice, 16),
+      gasLimit: new BN(paramsData.gasLimit, 16),
+      nonce: new BN(paramsData.nonce, 16),
+      hashLock: paramsData.hashLock,
+      messageHash: revokedProofData.gateway.stake.return_value.returned_value.messageHash_,
+      staker: paramsData.staker,
+      rlpParentNodes: revokedProofData.co_gateway.confirm_revert_stake_intent.proof_data.storageProof[0].serializedProof,
+      blockHeight: new BN(revokedProofData.co_gateway.confirm_revert_stake_intent.proof_data.block_number,16),
+      gatewayAddress: revokedProofData.contracts.gateway,
+      storageRoot: revokedProofData.co_gateway.confirm_revert_stake_intent.proof_data.storageHash,
+    };
+
+    await setup(revertStakeParams);
+
+    await mockToken.transfer(gateway.address, revertStakeParams.amount, { from: accounts[0] });
+    await baseToken.transfer(gateway.address, bountyAmount.muln(2.5), { from: accounts[0] });
+  });
+
+  it('should fail when message hash is zero', async function () {
+
+    await gateway.setStorageRoot(revertStakeParams.blockHeight, revertStakeParams.storageRoot);
+
+    await Utils.expectRevert(
+      gateway.progressRevertStake(
+        ZeroBytes,
+        revertStakeParams.blockHeight,
+        revertStakeParams.rlpParentNodes,
+      ),
+      'Message hash must not be zero.',
+    );
+
+  });
+
+  it('should fail when storage proof zero', async function () {
+
+    await gateway.setStorageRoot(revertStakeParams.blockHeight, revertStakeParams.storageRoot);
+
+    await Utils.expectRevert(
+      gateway.progressRevertStake(
+        revertStakeParams.messageHash,
+        revertStakeParams.blockHeight,
+        '0x',
+      ),
+      'RLP parent nodes must not be zero.',
+    );
+  });
+
+  it('should fail when storage proof is incorrect', async function () {
+
+    let blockHeight = new BN(revokedProofData.co_gateway.confirm_stake_intent.proof_data.block_number,16);
+    let storageRoot = revokedProofData.co_gateway.confirm_stake_intent.proof_data.storageHash;
+
+    await gateway.setStorageRoot(blockHeight, storageRoot);
+
+    await Utils.expectRevert(
+      gateway.progressRevertStake(
+        revertStakeParams.messageHash,
+        blockHeight,
+        revokedProofData.co_gateway.confirm_stake_intent.proof_data.storageProof[0].serializedProof,
+      ),
+      'Merkle proof verification failed.',
+    );
+
+  });
+
+  it('should fail when storage proof is invalid', async function () {
```

Comment:
I am not sure if we need this test. This seems like testing of MPP library.

Reply:
It is not a pure unit test, its an integration of Gateway+GatewayLib+MessageBus+MPP as a unit.

Reply:
Integration of multiple units is generally not a unit, but an integration 🤔

Reply:
Yes, I understand this but we decided not to replicate code with `MockMessageBus` approach,  so all the progress with the proof tests is integration w.r.t `messageBus`, `gatewayLib` and `MPP`.

In such a scenario, I am not sure where to draw a hard line for what to test and not to. So I add the tests considering all as a unit. 

I am in favor to keep such tests where ever we have integration. But if the majority of us decides that it is not needed then we can remove it.